### ISSUE TITLE
Cosine decay schedule and restarts

### ISSUE THREAD

**Issue by tisabe** [google-deepmind/optax#304]
In the [paper](https://arxiv.org/abs/1608.03983) linked with `optax.cosine_decay_schedule` the key is the restarting of the learning rate, i.e. the learning rate having a cosine shape. But in the optax implementation there is only one half period and then the learning rate is zero.

If this is intended, it should be mentioned in the documentation and a different periodic schedule could be made available, or correct me if my understanding of the paper or implementation is wrong.

**Comment by samuela** [comment]
Just checked and was shocked to find the same thing. It looks like it's only doing a single half-period, and then it's just learning rate zero. This is a bug, no? How does one implement the cosine decay schedule from the paper?

**Comment by samuela** [comment]
Just talked with a friend that works in CV... he reports that doing a single half-period decay is actually standard these days. So I guess there's some precedent for this.

**Comment by mkunesch** [comment]
Hey! Thanks a lot for raising this!

For `cosine_decay_schedule` I think this is the desired behaviour in the user codes I looked at (the learning rate stops at `alpha * init_value` so if `alpha` is not zero it does not decay all the way to zero). In the discussion, the paper also advocates the use of the cosine schedule even without restarts.

However, I totally agree that all this should be explained in the docstring - since the paper is cited and it is about restarts, it is confusing that there are no restarts in the schedule.

I'm not sure whether implementing a separate schedule for restarts with cosine decay is the best approach: if I understand the paper correctly, the restarts should be done with ever increasing `decay_steps`. In practice, I think this means that it would be easiest to use `optax.join_schedules` to join several different versions of `cosine_decay_schedule` together.

I've file PR #310 to clarify the docstring and mention `optax.join_schedules`. Let me know what you think!

Thanks again!

**Comment by mkunesch** [comment]
Actually, it would be great to have an example that shows the use of schedules, e.g. with `join_schedules` to get a linear increase at the beginning, followed by cosine decay with several restarts.

Would you like to write this example? If not, I'll create a new issue requesting this example and label it as "good first issue".

**Comment by samuela** [comment]
Thanks for taking a look at this and clarifying the docs in #310, @mkunesch! I tried using something like
```py
  lr_schedule = optax.join_schedules(schedules=[
      optax.warmup_cosine_decay_schedule(
          init_value=1e-6,
          peak_value=learning_rate,
          warmup_steps=warmup_epochs * steps_per_epoch,
          decay_steps=(cycle_length - warmup_epochs) * steps_per_epoch)
  ] * num_cycles,
                                     boundaries=jnp.cumsum(jnp.array([cycle_length] * num_cycles)))
```
but found that it didn't do much to help me (training VGG16 on CIFAR-10). It's also prob not quite what was described in the original paper. I also tried decaying the peak learning rate exponentially with each cycle, but that gave me terrible performance for some reason I have yet to understand.

**Comment by tisabe** [comment]
Thanks for clarifying in the docstring and the discussion of this issue!

IMO it would be nice to have an example showing a schedule like @samuela suggested, but it would be better if it also increases performance compared to a standard schedule. Reproducing the results in the original paper with restarts would be ideal.

At the moment I unfortunately don't really have the time to write an example myself.
If the restarting schedule is not state of the art, it's maybe not that important right now.

**Comment by ShirongLiu** [comment]
how can i  use `optax.join_schedules` and `optax.cosine_decay_schedule` to implement `torch.optim.lr_scheduler.CosineAnnealingLR` in a simple way?
